src/modules/addimage.js

```diff
@@ -730,20 +734,28 @@ import { atob, btoa } from "../libs/AtobBtoa.js";
   var arrayBufferToBinaryString = (jsPDFAPI.__addimage__.arrayBufferToBinaryString = function(
     buffer
   ) {
-    try {
-      return atob(btoa(String.fromCharCode.apply(null, buffer)));
-    } catch (e) {
-      if (
-        typeof Uint8Array !== "undefined" &&
-        typeof Uint8Array.prototype.reduce !== "undefined"
-      ) {
-        return new Uint8Array(buffer)
-          .reduce(function(data, byte) {
-            return data.push(String.fromCharCode(byte)), data;
-          }, [])
-          .join("");
+    // Skip direct conversion as it might take many hundred milliseconds before throwing.
+    if (buffer.length < ARRAY_BUFFER_LIMIT) {
+      try {
+        return atob(btoa(String.fromCharCode.apply(null, buffer)));
+      } catch (e) {
+        // Buffer was overflown, fall back to other methods
       }
     }
+    // Text decoder is much faster than string concatenation or reduce/join.
+    if (typeof TextDecoder !== "undefined") {
+      var decoder = new TextDecoder("ascii");
+      return decoder.decode(buffer);
+    } else if (
+      typeof Uint8Array !== "undefined" &&
+      typeof Uint8Array.prototype.reduce !== "undefined"
+    ) {
+      return new Uint8Array(buffer)
+        .reduce(function(data, byte) {
+          return data.push(String.fromCharCode(byte)), data;
+        }, [])
+        .join("");
+    }
```

Comment:
In JavaScript, functional programming is not very efficient. I suggest removing `.reduce()` and replacing it with a simple for loop with string growing, since that is actually faster most of the the time (I've done extensive testing when building `fflate`, the compression library used in `jsPDF`). Also, all browsers with `ArrayBuffer` support also have `Uint8Array` support, so no need to check.
```suggestion
    } else {
      var out = "";
      var u8buf = new Uint8Array(buffer);
      for (var i = 0; i < u8buf.length; i += 16384) {
        out += String.fromCharCode.apply(null, u8buf.subarray(i, i + 16384));
      }
      return out;
    }
```

Actually, you could probably replace *all three* cases with this suggestion, since the second doesn't work and the first is around same performance (it is slightly better in modern browsers).

Reply:
Yeah, after some testing this is the best solution.